### DISCUSSION TITLE
Potential bug when using $fetch? regular fetch working but $fetch not

### DISCUSSION THREAD

**Discussion by ahsan-a** [nuxt/nuxt#15900] [Questions] [answered]
I've been trying do implement discord oauth2 into a new nuxt 3 application. For my token request i have done this:
```ts
	const response: any = await $fetch('https://discord.com/api/oauth2/token', {
		method: 'POST',
		headers: {
			'content-type': 'application/x-www-form-urlencoded',
		},
		body: new URLSearchParams({
			client_id: process.env.DISCORD_CLIENT_ID,
			client_secret: process.env.DISCORD_CLIENT_SECRET,
			grant_type: 'authorization_code',
			redirect_uri: 'http://localhost:3000/api/oauth/callback',
			code: code.toString(),
		}).toString(),
	});
```
When using this, i get this error on the api route (really nice looking error page btw):
![image](https://user-images.githubusercontent.com/50714675/151406855-7ab8f871-fe72-4330-8ddc-8965afec08a3.png)

and this in my console:
```
401 Unauthorized (https://discord.com/api/oauth2/token)
  at processTicksAndRejections (node:internal/process/task_queues:96:5)  
  at async callback (file://./.nuxt/nitro/index.mjs:157:20)  
  at async handle (file://./node_modules/h3/dist/index.mjs:601:19)
```

However, when removing the dollar sign and using regular fetch it seems to work completely fine, and i get the token successfully:
```ts
	const response: any = await fetch('https://discord.com/api/oauth2/token', {
		method: 'POST',
		headers: {
			'content-type': 'application/x-www-form-urlencoded',
		},
		body: new URLSearchParams({
			client_id: process.env.DISCORD_CLIENT_ID,
			client_secret: process.env.DISCORD_CLIENT_SECRET,
			grant_type: 'authorization_code',
			redirect_uri: 'http://localhost:3000/api/oauth/callback',
			code: code.toString(),
		}).toString(),
	}).then((x) => x.json());
```

I've opened a discussion instead of an issue as i'm not sure whether i've implemented this correctly. Are there any changes that need to be made to the above implementation in order for this to work as it does with `fetch`?

**Comment by folamy** [comment]
Try importing $fetch from ohmyfetch
[Ohmyfetch](https://github.com/unjs/ohmyfetch]

**Reply by ahsan-a** [reply]
The 401 error still occurs, is it likely an issue with my code and the package instead of it and nuxt?

**Comment by wobsoriano** [comment, marked as answer]
Where are you doing this? In an API route? If yes, get the config via [runtime config](https://v3.nuxtjs.org/docs/usage/runtime-config)

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  privateRuntimeConfig: {
    DISCORD_CLIENT_ID: process.env.DISCORD_CLIENT_ID,
    DISCORD_CLIENT_SECRET: process.env.DISCORD_CLIENT_SECRET
  }
})
```

Also, you can remove the header property from your request as [ohmyfetch](https://github.com/unjs/ohmyfetch/commit/4adb3bc38d9423507ccdcd895b62a3a7d95f4144) automatically checks the body and stringifies it

```ts
// server/api/discord.ts
import config from '#config'

export default async (req, res) => {
    const response: any = await $fetch('https://discord.com/api/oauth2/token', {
		method: 'POST',
		body: new URLSearchParams({
			client_id: config.DISCORD_CLIENT_ID,
			client_secret: config.DISCORD_CLIENT_SECRET,
			grant_type: 'authorization_code',
			redirect_uri: 'http://localhost:3000/api/oauth/callback',
			code: code.toString(),
		}),
	});

    return {}
}
```

**Reply by ahsan-a** [reply]
This worked, and then it suddenly stopped working. i'm not sure why but when replacing for fetch again, it works.

**Reply by wobsoriano** [reply]
yeah would be nice to see a minimal code repro so we can check

**Reply by ahsan-a** [reply]
https://codesandbox.io/s/recursing-sun-ezmcj?file=/server/api/callback.ts
here, i've made a new testing discord app so i'm fine with the client secret being leaked, because it uses oauth you'll need to have discord. At the beginning this didn't work, but it suddenly started working and i'm not sure how.

https://ezmcj.sse.codesandbox.io/

**Reply by guillaumeprevost** [reply]
Hi ! I encountered the same issue : a simple call using $fetch would not work (hence why I end up on this thread). 

```
const params = {
    email: auth?.user?.email,
    password: confirmationCode.value,
    grant_type: 'password',
  };
  const urlParams = new URLSearchParams(params);
  $fetch( `${config.apiBaseUrl}/api/oauth/token?${urlParams.toString()}`, {
    method: 'POST',
    headers: {
      Authorization: authHeader,
      "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
    },
  })
  .then(async (authToken) => {
    ...
  });
```

Switching to fetch suddenly made it work without any other change in my code.

**Reply by florianschepp** [reply]
https://github.com/unjs/ofetch#%EF%B8%8F-parsing-response

> For binary content types, ofetch will instead return a Blob object.

I found this statement not to be entirely true.

Looking at ofetch and how it defines the responseType:
```
const JSON_RE = /^application\/(?:[\w!#$%&*.^`~-]*\+)?json(;.+)?$/i;

export function detectResponseType (_contentType = ""): ResponseType {
  if (!_contentType) {
    return "json";
  }

  // Value might look like: `application/json; charset=utf-8`
  const contentType = _contentType.split(";").shift()!;

  if (JSON_RE.test(contentType)) {
    return "json";
  }

  // TODO
  // if (contentType === 'application/octet-stream') {
  //   return 'stream'
  // }

  if (textTypes.has(contentType) || contentType.startsWith("text/")) {
    return "text";
  }

  return "blob";
}
```
you'll see that for the contentType here then being x-www-form-urlencoded it defaults to blob as well. Luckily for now you can force a JSON response by doing 
`await ofetch('/movie?lang=en', { parseResponse: JSON.parse })`
as stated in the above linked documentation.
